Added support for LDAP authentication

maheshkelkar opened the pull request:
This can be used for authentication required for manual promotion or for all
REST apis in future

- Added LDAP client support using JNDI.
- Using dropwizard basic HTTP authenticator support for fetching the attributes from client.
- Added LDAP test server using unboundid's in-memory ldap server.
- Added spock and cucumber tests

References #157

maheshkelkar commented:
@GiriDandu Please review

rtyler commented:
I'll take a look after travis green lights

maheshkelkar commented:
@rtyler @GiriDandu Travis is happy now. The changes are ready for review

GiriDandu commented on build.gradle:
```diff
@@ -59,6 +60,7 @@ dependencies {
     testCompile 'cglib:cglib-nodep:2.2.+'
 
     cucumberCompile 'info.cukes:cucumber-groovy:1.2.+'
+    cucumberCompile 'com.unboundid:unboundid-ldapsdk:2.3.8'
```
Why add only ldap only for compile time? How about when we run in shadowJar or real binary? How will this work?

maheshkelkar replied:
This is for test LDAP server. As for JNDI is already part of JAVA package

rtyler replied:
Comments like this are useful to put into the file itself for posterity FWIW

GiriDandu commented on build.gradle:
```diff
@@ -70,7 +72,6 @@ dependencies {
 repositories {
     jcenter()
     mavenCentral()
-
```
Remove this space

maheshkelkar replied:
No space, I have removed whole line.

GiriDandu commented on deploydb.h2.cucumber.yml:
```diff
@@ -22,4 +22,16 @@ logging:
       currentLogFilename: ./logs/deploydb-cucumber.log
       threshold: ALL
       archive: false
-      timeZone: UTC
```
Remove space

maheshkelkar replied:
Actually, there is no space there. I have added newline (to add ldap config) which was missing earlier

GiriDandu commented on src/asciidoc/auth.ad:
```diff
@@ -1,6 +1,6 @@
 = #DeployDB Authentication
 
-== Basic Authentication for Administrator
+== Basic HTTP Authentication for Administrator
```
We talked about Authentication not required for the administrator in the last meeting. Is this still need?

maheshkelkar replied:
I have not added anything, but borrowed this from dropwizard. Simply idenitifying what user can do for deploydb. No code changes for this in deploydb.

GiriDandu commented on src/asciidoc/auth.ad:
```diff
@@ -12,29 +12,27 @@ http:
 ----
 ldap:
-  uri: ldaps://myldap.com:636
+  uri: ldap://server.lookout.com:10389
```
Please remove the lookout name from the documentation.  Make this "server.example.com"

maheshkelkar replied:
done

GiriDandu commented on src/cucumber/groovy/deploydb/cucumber/AppHelper.groovy:
```diff
@@ -145,9 +148,11 @@ class AppHelper {
+        return getClient().target(urlWithPort(path, isAdmin))
                      .request()
                      .build(method, entity)
+                     .property(HttpAuthenticationFeature.HTTP_AUTHENTICATION_BASIC_USERNAME, "peter")
```
Who is "peter" :-). Can this added to deploydb.h2.cucumber.yml ? Remove the hard coding. Same for "griffin" also

maheshkelkar replied:
we cannot put this in launch config. This is part of the "step". Let me think about the solution. I will take care of this.

GiriDandu commented on src/cucumber/groovy/deploydb/cucumber/TestLdapServer.groovy:
```diff
@@ -0,0 +1,89 @@
+import org.slf4j.LoggerFactory
+
+
+final class TestLdapServer {
```
Can you move this out of cucumber directory. In future we may want to test with spock and it will be easy integration to do this.

maheshkelkar replied:
I am waiting for your changes to complete. Once you are ready, we can simply move 1 file to whichever common directory you have. So if I merge before, you may have to merge it.

GiriDandu commented on src/cucumber/groovy/step_definitions/HttpSteps.groovy:
```diff
@@ -67,7 +68,10 @@ When(~/^I GET "(.*?)" with custom headers:$/) { String path, DataTable headers -
 
-    response = builder.build('GET', null).invoke()
+    response = builder.build('GET', null)
+            .property(HttpAuthenticationFeature.HTTP_AUTHENTICATION_BASIC_USERNAME, "peter")
```
Make "peter" and "griffin" configurable

maheshkelkar replied:
will take care of it

GiriDandu commented on src/main/groovy/deploydb/DeployDBApp.groovy:
```diff
@@ -167,6 +171,19 @@ class DeployDBApp extends Application<DeployDBConfiguration> {
         environment.admin().addTask(new ConfigReloadTask(workFlow))
 
+        /** Register Ldap Authentication */
+        auth.LdapConfiguration ldapConfiguration = configuration.getLdapConfiguration()
```
If the ldap is not present in the yml file, is the ldapConfiguration object is created? may be you want to check the object before calling uri null check

maheshkelkar replied:
LdapConfiguration is always present with default parameters

GiriDandu commented on src/test/groovy/deploydb/auth/LdapAuthenticatorSpec.groovy:
```diff
@@ -0,0 +1,232 @@
+
+        then:
+        thrown(io.dropwizard.auth.AuthenticationException)
+    }
```
:+1:

rtyler commented on src/main/groovy/deploydb/auth/LdapAuthenticator.groovy:
```diff
@@ -0,0 +1,135 @@
+     * @throws NamingException
+     */
+    InitialDirContext buildContext(String sanitizedUsername, String password) throws NamingException {
+        final String userDN = String.format("%s=%s,%s", configuration.getUserNameAttribute(),
```
IMO this string formatting code should be separated into a separate well tested function instead of putting that logic into this method

maheshkelkar replied:
Not sure what you mean by separating this formatting? Do you want to make sure that patterns are created correctly or arguments have valid values or if server is interpreting it correctly?

rtyler replied:
This is logic that should be put into its own method so it can be tested in isolation.

rtyler commented on src/main/groovy/deploydb/auth/LdapAuthenticator.groovy:
```diff
@@ -0,0 +1,135 @@
+     * @return Directory context
+     * @throws NamingException
+     */
+    InitialDirContext buildContext(String sanitizedUsername, String password) throws NamingException {
```
Lots of questions about this function which should be addressed both in updated javadocs and comments in this PR:

 1. What happens for invalid URIs?
 1. What happens if the connection or read timeout values are exceeded?
 1. Is pooling of LDAP connections handled underneath the covers here?
 1. What if configuration is null?

rtyler commented on src/main/groovy/deploydb/auth/LdapAuthenticator.groovy:
```diff
@@ -0,0 +1,135 @@
+        final Hashtable<String, String> env = new Hashtable<>()
+
+        /** Common attributes */
+        env.put(Context.INITIAL_CONTEXT_FACTORY, "com.sun.jndi.ldap.LdapCtxFactory")
```
All of these strings should be constants on the class

rtyler commented on src/main/groovy/deploydb/auth/LdapAuthenticator.groovy:
```diff
@@ -0,0 +1,135 @@
+        /** Filter the user groups for configured groupClass (aka objectClass) */
+        final String filter = String.format("(&(%s=%s)(objectClass=%s))",
+                configuration.getGroupMembershipAttribute(), userName,
+                configuration.getGroupClassName(        ))
```
`.getGroupClassName( ))` why so much space o_O

rtyler commented on src/main/groovy/deploydb/auth/LdapAuthenticator.groovy:
```diff
@@ -0,0 +1,135 @@
+    private Set<String> getGroupMemberships(InitialDirContext context, String userName) {
+
+        /** Filter the user groups for configured groupClass (aka objectClass) */
+        final String filter = String.format("(&(%s=%s)(objectClass=%s))",
```
This string formatting should be refactored into a well tested helper function

rtyler commented on src/main/groovy/deploydb/auth/LdapAuthenticator.groovy:
```diff
@@ -0,0 +1,135 @@
+
+        /** Limit the search results to groupNameAttribute only */
+        SearchControls searchCtls = new SearchControls();
+        def returnedAtts = [configuration.getGroupNameAttribute()] as String[]
```
Why isn't `returnedAtts` typed? e.g. `String[] returnedAttrs`

rtyler commented on src/main/groovy/deploydb/auth/LdapAuthenticator.groovy:
```diff
@@ -0,0 +1,135 @@
+                configuration.getGroupClassName(        ))
+
+        /** Limit the search results to groupNameAttribute only */
+        SearchControls searchCtls = new SearchControls();
```
No need for the semicolon

rtyler commented on src/main/groovy/deploydb/auth/LdapAuthenticator.groovy:
```diff
@@ -0,0 +1,135 @@
+import org.slf4j.LoggerFactory
+
+
+class LdapAuthenticator implements Authenticator<BasicCredentials, BasicCredentials> {
```
I think this class should be annotated with `@TypeChecked` to help ensure we're doing the right things here

rtyler commented on src/main/groovy/deploydb/auth/LdapAuthenticator.groovy:
```diff
@@ -0,0 +1,135 @@
+     */
+    @Override
+    Optional<User> authenticate(BasicCredentials credentials) throws AuthenticationException {
+        String pass = credentials.getPassword()
```
I'd prefer to un-Java this code a bit and use `credentials.password` and `credentials.username` instead of these getter functions :P

maheshkelkar replied:
done

rtyler commented on src/main/groovy/deploydb/auth/LdapAuthenticator.groovy:
```diff
@@ -0,0 +1,135 @@
+    Optional<User> authenticate(BasicCredentials credentials) throws AuthenticationException {
+        String pass = credentials.getPassword()
+
+        final String sanitizedUsername = sanitizeEntity(credentials.getUsername())
```
Is there a compelling reason to sanitize the username as opposed to throwing an `InvalidUserNameException` of some form?  Trimming leading or trailing space I think is fine, but I'm worried about mutating user input for authentication purposes.

maheshkelkar replied:
I have removed this part. Because if username is invalid, I assume LDAP authentication should fail and complain about.

rtyler commented on src/main/groovy/deploydb/auth/LdapAuthenticator.groovy:
```diff
@@ -0,0 +1,135 @@
+        } catch (AuthenticationException ae) {
+            logger.error("${sanitizedUsername} failed to authenticate: ", ae)
+        } catch (NamingException err) {
+            throw new io.dropwizard.auth.AuthenticationException(
```
Why is this `Authentication` using the full classpath? this can be shortened up

maheshkelkar replied:
JNDI throws javax.naming.AuthenticationException and this code is in-turn throwing io.dropwizard.auth.AuthenticationException, which dropwizard internals can handle and interpret

rtyler replied:
I see, well,  that's confusing :)

rtyler commented on src/main/groovy/deploydb/auth/LdapConfiguration.groovy:
```diff
@@ -0,0 +1,123 @@
+        return uri
+    }
+
+    LdapConfiguration setUri(URI uri) {
```
Why does this file have all these getter and setters? I don't understand why they're necessary

maheshkelkar replied:
done, taken care of

rtyler commented on src/main/groovy/deploydb/auth/User.groovy:
```diff
@@ -0,0 +1,14 @@
+package deploydb.auth
+
+/**
+ * User class
```
This needs a bit more description, why do we have our own User class, etc.

maheshkelkar replied:
done

rtyler commented on src/main/groovy/deploydb/resources/DeploymentResource.groovy:
```diff
@@ -29,6 +27,7 @@ import javax.ws.rs.WebApplicationException
 import javax.ws.rs.core.Context
 import javax.ws.rs.core.MediaType
 import javax.ws.rs.core.Response
+import org.apache.commons.lang3.tuple.Pair
```
Pair doesn't appear to be used so it should be removed

maheshkelkar replied:
It is used.

rtyler replied:
AH, sorry, I misread this earlier and didn't notice that this was just moving further down in the import list for some reason

rtyler commented on src/test/groovy/deploydb/auth/LdapAuthenticatorSpec.groovy:
```diff
@@ -0,0 +1,232 @@
+        0 * context._
+
+        /* buildContext partial mocking */
+        ldapAuthenticator.metaClass.buildContext = { String username, String pass -> return context }
```
This is *not* how mocking/stubbing should be done with Spock. Please review the Spock documentation for how to mock and stub properly.

rtyler commented on src/test/groovy/deploydb/auth/LdapAuthenticatorSpec.groovy:
```diff
@@ -0,0 +1,232 @@
+
+    def "Successful authentication and getGroupMembership"() {
+        given:
+        def ldapConfiguration = new auth.LdapConfiguration()
```
Why aren't these variables typed? AFAICT we don't gain anything from not being typed

rtyler commented on src/test/groovy/deploydb/auth/LdapAuthenticatorSpec.groovy:
```diff
@@ -0,0 +1,232 @@
+        matchAttrs.put(new BasicAttribute("app", "Fandango")); /* Does not match groupNameAtribute */
+        def result = new SearchResult("fauxname", null, matchAttrs)
+
+        /* Context calls mocking */
```
The number of test cases that have this mocking repeated is a good indication that these functions need to be refactored out a *lot* more than they already are so you can more granularly write tests without needing to mock out the world for interacting with LDAP

rtyler commented:
@maheshkelkar many of my comments in the spock tests apply to all the added spock test cases, not just the ones I commented on.

maheshkelkar commented:
@rtyler @GiriDandu I have tried to incorporate all of your comments. I have simplified the LdapAuthenticator, added several javadocs to fill in details, simplified spock tests to make them more readable. 

Please review when Travis is green and let me know your thoughts and comments.

GiriDandu commented on features/api/deployment/reading.feature:
```diff
@@ -36,6 +36,13 @@ Feature: Deployment READ APIs
       }]
     """
 
+  @error @wip
```
Unless you will take care of next commit/PR, can you please add an issue for this, that way you don't forget to cleanup this later

GiriDandu commented:
:+1:

rtyler commented:
There's still a number of changes that I wanted to see that didn't make it into this revision. That said,  this change is too fucking big to review sanely any more, so I'll merge it and start filing clean up tasks tomorrow
